composer/utils/checkpoint.py

```diff
@@ -217,17 +217,33 @@ def read_data(self, plan: LoadPlan, planner: LoadPlanner):
         first_replica = self.device_mesh is None or self.device_mesh.ndim == 1 or (
             self.device_mesh.ndim >= 2 and self.device_mesh.get_local_rank(mesh_dim=0) == 0)
 
-        # 1. Download to the destination all files this rank needs if on first replica
+        # 1. Collect the relative paths to download for all ranks for deduplication
+        relative_file_paths = set()
+        for plan_item in plan.items:
+            relative_file_paths.add(self.storage_data[plan_item.storage_index].relative_path)
+        all_file_paths = dist.all_gather_object(relative_file_paths)
+        log.debug(f'Files required for all ranks: {all_file_paths}.')
+
+        # 2. Download to the destination all files this rank needs if on first replica
         if first_replica:
             log.debug(f'Rank {dist.get_global_rank()} starting to download files.')
+
+            # Get the lowest rank in the current node
+            local_rank_0 = dist.get_global_rank() // dist.get_local_world_size() * dist.get_local_world_size()
+
             for plan_item in plan.items:
-                # Each plan item has a storage index which points to the relative path of the shard file at save time.
                 relative_file_path = self.storage_data[plan_item.storage_index].relative_path
+                # Check if the file is scheduled to be downloaded by a lower rank on the same node
+                # i.e. if rank 9 and rank 10 on the same node have the same the same required file, rank 10 should not download it.
```

Comment:
```suggestion
                # i.e. if rank 0 and rank 1 on the same node have the same the same required file,
                # only rank 0 should download it and not rank 1.
```
Let's keep example to node 0 for code since it's easier to understand. I gave a rank 1 example for Evan to imply multinode, but for comments likely more confusing